### PR TITLE
feat: gRPC interface for ingester->querier v2

### LINKED ISSUE #8346: Ingester->Querier v2: Design gRPC protocol
Design gRPC protocol for #8169.

Rough idea:

```protobuf
message IngesterQueryRequest {
  // Namespace to search
  int64 namespace_id = 1;

  // Table that should be queried.
  int64 table_id = 2;

  // Projection expressed as column IDs
  repeated int64 projection = 3;

  // Predicate for filtering
  optional Predicate predicate = 4;
}

message IngesterQueryReponseMetadata {
  message Partition {
    // Partition ID.
    int64 id = 1;

    // Minimum timestamp.
    int64 t_min = 2;

    // Maximum timestamp (inclusive).
    int64 t_max = 3;

    // Projection in terms of column IDs.
    repeated int64 projection = 4;
  }

  uuid ingester_uuid = 1;
  int64 persist_counter = 2;
  bytes table_schema = 3;
  repeated Partition partitions = 4;
}

message IngesterQueryReponsePayload {
  // Partition ID.
  int64 partition_id = 1;

  // Projection in terms of column IDs.
  //
  // This MUST be a subset of the projection transmitted within the metadata
  repeated int64 projection = 2;

  // Serialized RecordBatch (w/o schema)
  bytes record_batch = 3;
}

message IngesterQueryResponse {
  oneof msg {
    // Metadata, this is ALWAYS the first message and MUST NOT be repeated.
    IngesterQueryReponseMetadata metadata = 1;

    // Payload, following the first message.
    IngesterQueryReponsePayload payload = 2;
  }
}

service IngesterQueryService {
    rpc query (IngesterQueryRequest) returns (stream IngesterQueryResponse);
}
```

This might just go go `generated_types`.

### PR DESCRIPTION
**Pull request by crepererum**
Note that the actual "to/from bytes" conversion for the schema and batches will be implemented in #8347.

Closes #8346.

### CODE REVIEW

File: generated_types/protos/influxdata/iox/ingester/v1/query.proto [hunk 1]
```diff
@@ -0,0 +1,75 @@
+syntax = "proto3";
```
Comment (alamb) [inline generated_types/protos/influxdata/iox/ingester/v1/query.proto:1]: There appears to already be a Ingester / Querier gRPC proto file here: 

https://github.com/influxdata/influxdb_iox/blob/main/ingester_query_grpc/protos/influxdata/iox/ingester/v1/query.proto

Perhaps we should add the new requests / responses to the same location to keep things easier to discover?

It appears to have been done in https://github.com/influxdata/influxdb_iox/pull/7789 -- Perhaps @carols10cents  has more thoughts on the matter

Comment (carols10cents) [reply generated_types/protos/influxdata/iox/ingester/v1/query.proto:1]: The `generated_types` crate is for *external* gRPCs; the `ingester_query_grpc` crate is for *internal* gRPC APIs that external clients don't need to know about/pull in the dependencies for.

> why do we have two crates for this now?

Answered in the PR Andrew linked to: https://github.com/influxdata/influxdb_iox/pull/7789#discussion_r1193843684

File: ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  repeated string columns = 3;
+
+  // Predicate for filtering.
+  optional Predicate predicate = 4;
+}
+
+message IngesterQueryReponseMetadata {
+  message Partition {
+    // Partition ID.
+    int64 id = 1;
```
Comment (carols10cents) [inline ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto:32]: This needs to be `oneof` `catalog_id` or `hash_id` now, like it is in v1: https://github.com/influxdata/influxdb_iox/blob/7a462754e4123c76244ec9df571e6d923cbd06d2/ingester_query_grpc/protos/influxdata/iox/ingester/v1/query.proto#L83-L104

File: ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  // Serialized table schema.
+  bytes table_schema = 3;
+
+  // Ingester partitions.
+  repeated Partition partitions = 4;
+}
+
+message IngesterQueryReponsePayload {
+  // Partition ID.
+  int64 partition_id = 1;
```
Comment (carols10cents) [inline ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto:62]: And here needs to be `oneof` `catalog_id` or `hash_id` too

File: ingester_query_grpc/src/lib.rs [hunk 4]
```diff
@@ -214,6 +265,49 @@ impl TryFrom<Predicate> for proto::Predicate {
...
+
+    fn try_from(pred: Predicate) -> Result<Self, Self::Error> {
+        let Predicate {
+            field_columns,
+            range,
+            exprs,
+            value_expr,
+        } = pred;
+
+        if field_columns.is_some() {
```
Comment (carols10cents) [inline ingester_query_grpc/src/lib.rs:279]: Is it ever possible that the querier would send `Some([])` (`Some` empty set) for `field_columns`? I suppose it's our own fault if it does. 

I wonder if these checks are pointing to a need for a new `Predicate` type that doesn't have `field_columns`, `range`, `value_expr` at all?

**Review by carols10cents — CHANGES_REQUESTED** [review]
Two spots that need to be updated to reflect recent changes to partition identification that v2 will need as well; one question that I'm not sure matters or not.

File: ingester_query_grpc/src/lib.rs (hunk 4 above)
Comment (crepererum) [reply ingester_query_grpc/src/lib.rs:279]: yeah, we should probably not use `Predicate` here AT ALL but `Vec<Expr>` directly. `Predicate` is a leftover from the InfluxRPC storage API and I plan to remove it from the querier internals anyways: #8097. I'll adjust the gRPC interface...

Comment (crepererum) [reply ingester_query_grpc/src/lib.rs]: done

File: ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto (hunk 3 above)
Comment (crepererum) [reply ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto]: done

File: ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto (hunk 2 above)
Comment (crepererum) [reply ingester_query_grpc/protos/influxdata/iox/ingester/v2/query.proto]: done

**Review by carols10cents — APPROVED** [review]
A few tiny typos I trust you to resolve. Much better though, thanks!

**Pull request merged**